azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/MessagingFactory.java

```diff
@@ -422,24 +422,51 @@ public void onError(Exception error) {
 
     private class RunReactor implements Runnable {
         final private Reactor rctr;
+        final private Executor executor;
 
-        public RunReactor(final Reactor reactor) {
+        volatile boolean hasStarted;
+
+        public RunReactor(final Reactor reactor, final Executor executor) {
             this.rctr = reactor;
+            this.executor = executor;
+            this.hasStarted = false;
         }
 
         public void run() {
-            if (TRACE_LOGGER.isInfoEnabled()) {
+            if (TRACE_LOGGER.isInfoEnabled() && !this.hasStarted) {
                 TRACE_LOGGER.info("messagingFactory[%s], hostName[%s], info[%s]",
                         getClientId(),
                         getHostName(),
                         "starting reactor instance.");
             }
 
+            boolean yieldedReactor = false;
+
             try {
-                this.rctr.setTimeout(3141);
-                this.rctr.start();
-                while (!Thread.interrupted() && this.rctr.process()) {
+                if (!this.hasStarted) {
+                    this.rctr.setTimeout(3141);
+                    this.rctr.start();
+                    this.hasStarted = true;
+                }
+
+                if (!Thread.interrupted() && this.rctr.process()) {
+                    try {
+                        this.executor.execute(this);
+                        yieldedReactor = true;
+                    } catch (RejectedExecutionException exception) {
+                        if (TRACE_LOGGER.isWarnEnabled()) {
```

Comment:
Let's make sure user will have a way to tell whether this issue happens because threadpool is closed by a user, or for some other reasons.

Reply:
will address this in the upcoming iteration...